agent/agent.go

```diff
@@ -943,6 +951,22 @@ func (a *agent) Close() error {
 	}
 	close(a.closed)
 	a.closeCancel()
+
+	rawMetadata := a.metadata.Load()
+	if rawMetadata == nil {
+		return xerrors.Errorf("no metadata was provided")
+	}
+	metadata, valid := rawMetadata.(codersdk.WorkspaceAgentMetadata)
+	if !valid {
+		return xerrors.Errorf("metadata is the wrong type: %T", metadata)
+	}
```

Comment:
We probably shouldn't do an early exit here and leave closure half-way after it's already started.